nimble/host/src/ble_hs_hci_evt.c

```diff
@@ -264,19 +256,14 @@ ble_hs_hci_evt_num_completed_pkts(uint8_t event_code, uint8_t *data, int len)
                     conn->bhc_outstanding_pkts -= num_pkts;
                 }
 
-                if (ble_hs_hci_avail_pkts == 0) {
```

Comment:
If there is no traffic on ACL, Nimble controller will send Number of Completed packets every X sec. I think it is 2 sec by default with num_pkts set to 0. I believe we don't need to call `ble_hs_wakeup_tx()` in such a case.

Maybe  instead of this change we could just remove `if (ble_hs_hci_avail_pkts == 0)` and it should be sufficient.

Reply:
I have not looked at the spec, but that is not how the nimble controller works.  The nimble controller only sends a 0-count num-completed-packets event if it has outgoing packets enqueued that it has not sent yet.  If the controller is idle, it does not send any num-completed-packets events (the code is a bit tricky; `ble_ll_conn_num_comp_pkts_event_send()`).

Like I said - I am not familiar with the controller sections of the spec.  Maybe the controller is allowed to send a num-completed-packets event every two seconds, even when idle.  I don't see that as a huge issue though.  Looping through some empty queues every two seconds in the host task should not present a performance problem.

Reply:
Here is the section from the spec:
While the Controller has HCI data packets in its buffer, it must keep sending the Number Of Completed Packets event to the Host at least periodically, until it finally reports that all the pending ACL Data Packets have been transmitted or flushed. The rate with which this event is sent is manufacturer specific.

As Chris says, the controller will not send 0 in its num completed packet event unless there is data enqueued. The idea behind not doing this periodically is that we did not want to allocate an event buffer for it if not needed.

I do not see why the controller would be prohibited from doing this but why should it? It is just extra work that does not need to be done (imo).

Reply:
@ccollins476ad @wes3   you are right, and change looks good.